### ISSUE TITLE
Where is the cache output?

### ISSUE THREAD

**Issue by fardarter** [microsoft/MSBuildCache#104]
I've followed the configuration instructions carefully, but I can't seem to find any kind of cache output?

What output am I supposed to cache? What location am I supposed to return it to for reuse between pipeline runs?

Really feeling challenged by this.

**Comment by fardarter** [comment]
@dfederm Hi, look, it may be in the docs, but that doesn't mean the docs have been well written to highlight it (especially for newcomers). For a project to be welcoming to non-experts, it really has to hold your hand in terms of getting an MVP working.

Can I suggest that it be used in a working example?

Also, why is it at the drive root and not the repo root?

What is the thinking behind this location? Is the intention for it to be a shared build cache for anything built?

How does the CI pipeline come to know where it is for caching? The CI example doesn't seem to be aware of it?

**Comment by dfederm** [comment]
> doesn't mean the docs have been well written to highlight it

Please feel empowered to improve the docs. I did the initial writing of the docs, but I certainly don't claim to be an expert technical writer. Contributions would be very welcome.

> Can I suggest that it be used in a working example?

That makes sense, and related to the above the docs probably do need a "getting started"-esque section. I'm open to suggestion for what a newcomer would like to see and would be helpful in understanding MSBuildCache.

> Also, why is it at the drive root and not the repo root? What is the thinking behind this location? Is the intention for it to be a shared build cache for anything built?

Mostly so that a git clean doesn't blow it away. Sharing across repos is nice, but isn't likely to provide a major benefit in practice.

The reason the location is configurable though is to allow for customization to suit your needs. The default value is informed by how we use this internally.

> How does the CI pipeline come to know where it is for caching? The CI example doesn't seem to be aware of it?

In general the local cache is not expected to be used by itself. Unless you have persistent build agents, it doesn't really help across builds.

It is however necessary to have a local cache so that within a single build each unique content file is only downloaded from the remote exactly once.

**Comment by fardarter** [comment]
> Please feel empowered to improve the docs. I did the initial writing of the docs, but I certainly don't claim to be an expert technical writer. Contributions would be very welcome.

If I get things working, I 100% am someone who can contribute that way. For now, I've abandoned the line of work, but if getting it set up becomes easy enough I'd pick it up again.

> In general the local cache is not expected to be used by itself. Unless you have persistent build agents, it doesn't really help across builds.

What I want to use is the Cache@2 Azure Devops task (see: https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/cache-v2?view=azure-pipelines). I don't like using persistent agents for a number of reasons, but cost and security are among them (security: so that there is very little pipelines/jobs/stages can do to interfere with each other; cost: so, I can use them on an on-demand basis).

> Mostly so that a git clean doesn't blow it away. Sharing across repos is nice, but isn't likely to provide a major benefit in practice.

That's fair enough. 

>  I'm open to suggestion for what a newcomer would like to see and would be helpful in understanding MSBuildCache.

So here is the problem I'm arrived with: I'd really like to be able to produce a deterministic hash/checksum from **source** files (multiple projects -- and signing would be post-build) with msbuild so that I can decide whether to build at all. The graph stuff here seemed about as close to anything I can find that would help me do that.

Failing that, I'd like to aggressively cache what I can. I've been vaguely thinking about caching intermediate objects (`obj`) as well. Not sure how sensible that is. What I imagined this project might be able to do is cache some sort of assessment of those objs so that msbuild is better able to interpret it. But honestly, I'm not really sure -- the docs seems to be written from 10,000 miles in the air rather than with clear and obvious problem statements for someone like me. Also open to caching in blob storage, but not sure what the difference would be, practically, between that and the ADO Cache Task.

Anyway, I wasn't able to quickly figure out where I should be getting an output, so I couldn't find anything to poke around in to verify I was doing sensible things or even understood what I could use this project for, exactly. And I'm not super keen to try figure out how to read binary logs.

So, I don't know if that perspective is helpful or informs the context I'm coming in with. If you know how to produce a prebuild hash of the source files using msbuild, I'd love a pointer in that direction. Otherwise, I'd love to know if I even understand what this project is for.

**Comment by dfederm** [comment]
> What I want to use is the Cache@2 Azure Devops task

Oh, then you should use the AzurePipelines implementation. It uses the same underlying cache backend as that `Cache@2` task. It should be more or less equivalent to using a local cache + `Cache@2`, but faster and properly scoped. If you're using Azure Pipelines, it should be really straightforward to just use this.

From what you've described, to get up and running all you really need to do is:
1. Add the packages as `GlobalPackageReference` items. This assumes you're using Central Package Management and are using projects which support PackageReferences (eg not C++ projects). Otherwise it's a slightly different process, but ultimately you just need to reference the `Microsoft.MSBuildCache.AzurePipelines` package.
2. Add `/graph /reportfileaccesses` to your MSBuild invocations in your pipeline.

See what that does. If there are errors you may need to address them as they come and how you address them depends on the errors and the behavior of your build. Many repos I've tested with have to do nothing at all, while others need to set a few properties to opt-out of some performance optimizations and/or strict correctness checks.

**Comment by fardarter** [comment]
> Oh, then you should use the AzurePipelines implementation. It uses the same underlying cache backend as that Cache@2 task. It should be more or less equivalent to using a local cache + Cache@2, but faster and properly scoped. If you're using Azure Pipelines, it should be really straightforward to just use this.

I followed all the docs for the ADO implementation, but I wasn't able to see any cache job running or make sense of where the files that I wanted to cache were or if they were being restored at all. My follow up assumption was that I needed to implement a cache task over some file location -- that is the output location I was thinking about when I started this issue.

If that's not the case, how can I see if it is even working? 

I have the same problem locally -- msbuild seemed to do a lot of things but I couldn't see any results or understand how to understand if I'd succeeded. Builds didn't appear dramatically faster -- I've subsequently marked certain projects as test projects, and I understand that may have interfered, but just having some clear sense/docs of what to expect when it's working, but locally and in CI would be super helpful.

> Add /graph /reportfileaccesses to your MSBuild invocations in your pipeline.

Side questions: 

- Is this still usable with `dotnet` cli and `msbuild` config properties? 
- If no, can one switch between `dotnet cli` and `msbuild` directly or do all stages need `msbuild`?

**Comment by dfederm** [comment]
So when using the Azure Pipelines implementation, it doesn't change the pipeline itself, ie there is no additional step in the pipeline. It uses the same underlying pipeline caching APIs, but it not a distinct step.

To evaluate whether the plugin is running or not, there are a few different things to look at.

Example when running locally for me when there are zero cache hits, so like what you'd expect on the very first build:
```
>msbuild /graph /reportfileaccesses /bl
MSBuild version 17.13.0-preview-24568-05+37fc8280d for .NET Framework
  Determining projects to restore...
  Restored <redacted> (in 7.67 sec).
  Restored <redacted> (in 9.92 sec).
  Restored <redacted> (in 21.01 sec).
Static graph loaded in 1.018 seconds: 3 nodes, 3 edges
Loading the following project cache plugin: MSBuildCacheLocalPlugin
  < ... build stuff ... >

Project cache statistics:
  Cache Hit Count: 0
  Cache Miss Count: 3
  Cache Hit Ratio: 0.0%

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:01:37.90
```

So you can see "Loading the following project cache plugin: MSBuildCacheLocalPlugin", as well as at the end some statistics.

After doing a `git clean -fdx` and then building again:

```
>msbuild /graph /reportfileaccesses /bl
MSBuild version 17.13.0-preview-24568-05+37fc8280d for .NET Framework
  Determining projects to restore...
  Restored <redacted> (in 2.53 sec).
  Restored <redacted> (in 3.31 sec).
  Restored <redacted> (in 3.32 sec).
Static graph loaded in 1.018 seconds: 3 nodes, 3 edges
Loading the following project cache plugin: MSBuildCacheLocalPlugin
<redacted> -> Cache Hit
<redacted> -> Cache Hit

Project cache statistics:
  Cache Hit Count: 3 (saved 28.9 project-seconds)
  Cache Miss Count: 0
  Cache Hit Ratio: 100.0%

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:21.88
```

So with cache hits you see the same as before but with "Cache Hit" for cached projects as well as a difference in the statistics.

Additionally, in the binlog you'll see various cache plugin related things:

<img width="299" alt="Image" src="https://github.com/user-attachments/assets/0b4594ae-d5c8-4501-a01c-9b9e3014958c" />

Finally, in the repo root (unless otherwise redirected) there should be a `MSBuildCacheLogs` directory with various log files.

If you're not seeing any of this, the PackageReference might not be properly configured. If you are able to share a binlog, that would be helpful in debugging.

As far as where the cache output goes, it goes to wherever it would go to where those files are intended to go. Ie, if a project normally writes `src/Foo/bin/x64/Foo.dll` without caching, then on cache hit that file is written to that path. That's the whole point of the cache; the stat on cache hit is exactly the same as if there was no cache hit but it's faster.

> Is this still usable with dotnet cli and msbuild config properties?

No, unfortunately this is msbuild.exe only.

> If no, can one switch between dotnet cli and msbuild directly or do all stages need msbuild?

Yes, but as you might expect, only the `msbuild.exe` commands will get caching.

RE Central Package Management, that's not a dealbreaker at all. You just need to reference the package in whatever way is typical for your repo (eg PackageReference in a global file, etc.). It's not the intent in this repo's docs though to exhaustively enumerate all the ways to reference a NuGet package since these packages follow standard conventions.

> How would one go about doing that correctly? I'm using the artifacts layout as well.

Artifacts output layout is completely unrelated. That's how you've configured your outputs, nor the packages you consume. How you reference package is somewhat repo dependent. NuGet has many options. `GlobalPackageReference`, `PackageReference` in a global location, and `packages.config` are common ones. If your repo is public and you can link it, I can recommend how to do it for your specific scenario. But again, the packages are following the standard NuGet conventions, so you just need to reference them and consume their content in the common way.

> Is there really nothing in the dotnet world that does this?

No, there is no "built-in" mechanism in .NET for creating some kind of unique key or hash for the build.

> what is actually cached?

All project outputs, ie all file writes which happened as part of building the project. At least the file writes that happened somewhere under the repo root. Again the intent of caching is to fully replicate the disk state identically to if it didn't come from cache.

> I didn't actually realise this was the project/release version of this package

Can you suggest something I can do to clarify? The very first thing the "Usage" section of the docs explains is how to add the packages.

> I'd be very surprised if everything was OK

Well, it depends. The main things I've see which cause issues with caching are when projects write over other projects' outputs. That breaks the optimization of using hardlinks for cache hits, but you can just opt out of that optimization for those specific paths (or all paths if needed). There should be clear errors about that too if you hit this which refer to "duplicate outputs". Another common thing I've seen is if projects launch detached child processes which don't exit until *after* the project finishes building, as the observation of file reads and writes is incorrect. Most of the time with those cases though it's just telemetry or temp files, so can simply be excluded. Other tech debt like using .NET Framework instead of .NET (Core) don't impact hashing at all. I can certainly help you work through any issues you run into if you're able to (and allowed to) share logs.

**Comment by dfederm** [comment]
Re-reading this thread, I think you might be misunderstanding what the caching is doing.

This is *not* the same as `Cache@2` which is basically pointing at a folder and says "zip that up and throw it in a cache" and later you can restore that directory.

The intent of MSBuildCache is to replicate the outputs of a build, where the granularity is per MSBuild project. Building with cache hits or misses (or without caching at all) should leave the disk in the same state, just with varying levels of performance.

You can think of it similarly to how incremental builds work. If you build in a clean repo, it takes a while. Then you build again and its faster, but the end state on disk is (ideally) the same as before. "Caching" in this case is describing memoization of the build process.

**Comment by fardarter** [comment]
@dfederm I appreciate this longer write up and I'll take some time next week to try understand it properly. The repo isn't public, but I could provision you with a temp account at some stage, if you're interested. Would be willing to contribute docs back in exchange.

That aside, I do understand that this is a kind of incremental build cache, though not clear exactly how it works. Let me try express my expectation a bit more clearly:

Imagining this is done on a VM:

```
1. Provision VM ---> 2. Build 1 (slow) ---> 3. [some information stored about Build 1 + some build output written to disk] ---> 4. Build 2 (fast -- consumes info + output).
```

Now imagine I want to get a fast build the first time, so what I imagine one could do is snapshot the disk after step 3.

```
Provision VM with snapshot from 3 above ---> Build 1 (fast, even if code has slightly changed).
```

To keep things simple, I'd like to not have to play with disk snapshotting for my runner hosts -- I want to use a managed devops pool with the azure runner images and a fresh (initial) state for each job (for the reasons given above). This means I use Cache tasks, artifacts and blob storage to manage state (with blob storage being the least desirable in terms of complexity re permissions, though it's a good option sometimes). What I'm trying to achieve is for pipeline caching and/or artifact production to play the role of the disk snapshot.

 ```
1. Provision runner Agent 1 ---> 2. Build 1 (slow) ---> 3. [some information stored about Build 1 + some build output written to disk] ---> 4. Build 2 (fast -- consumes info + output).
```

Then:

 ```
1. Provision fresh clean runner Agent 2 ---> 2. Acquire relevant data/output/info from step 3 above ---> 3. Build 2 (fast -- consumes info + output).
```

So ultimately the question is: What is the recipe for step 3? What combination of data/output/metadata/information could I restore to the new disk that would let agent 2 pretend, for the sake of caching the incremental build, that nothing has changed as we switched hosts?

Does that make sense? 

From an implementation perspective, I'm happy to cache quite a lot of stuff on a rolling basis (with a loose series of restore keys) with the assumption that the msbuild and this project is reasonably capable of understanding what it needs to rebuild when stuff is changed.

As an aside:

> No, there is no "built-in" mechanism in .NET for creating some kind of unique key or hash for the build.

![Image](https://github.com/user-attachments/assets/5c3884d9-320b-4266-865f-ae6660d32cc5)

I don't think I'll ever understand why this isn't a product priority for build tools. Slow CI is such a burden on devs and us devops people spend so much time and energy on designing pipelines. 

Is there a way to enumerate pre-build on one of the `msbuild` targets all the files that are going to be pulled into a build (per project or per solution; per project would be ideal)? Taking the `msbuild` view of what files are needed is going to be much more robust than any hack I put together with bash or powershell and git.

Anyway, understand if you don't want to indulge my curiosity there. The main question is the first one, and I'll do my best to get into the other details of your response next week.

**Comment by fardarter** [comment]
> Can you suggest something I can do to clarify? The very first thing the "Usage" section of the docs explains is how to add the packages.

It's a context knowledge thing. I'm a devops person who supports .NET teams, not a .NET dev, so decoding what's going on takes a bit of time and there's a lot to take in with `msbuild`. I'm not saying I was smart not to get on what was happening quickly, but I'm probably not the only person out there who would have read it poorly. I think I tried to see if it was declaring a cache version or whether it was an injected `msbuild` property of some kind, when it was a much simpler idea.

How about?

```xml
    <!-- Add your selected version of the MSBuildCache package, eg (v0.1.302-preview) -->
    <MSBuildCachePackageVersion>...</MSBuildCachePackageVersion>
```

**Comment by dfederm** [comment]
So I still don't think you're quite understanding MSBuildCache. This is _not_ "disk snapshotting".

Essentially, MSBuild's cache plugin asks a plugin if it wants to handle a build request just before it would execute it. This gives the plugin a chance to say "nah, I got this", and take over that build request for that project. ie, replay the file writes for that project from an equivalent previous build. So yes, it is laying files on disk, but it's _not_ the same as how you use the `Cache@2` task, or disk snapshotting.

Think of it like this. Let's say that I was asking you about capital cities. Some of them you might just have memorized so you can answer me very quickly. But in other cases, you might not know the answer, so we have to go look it up. But then later if I or someone else ask for that same city again, you might remember the one we looked up and be able to respond. That's memoization in a nutshell, "remembering" results which are expensive to compute. The equivalent to disk snapshotting in this analogy would be if you write down on a piece of paper all capital cities you knew (or maybe the ones you've been asked before?) and handed that entire thing to me to use as reference.

This flexibility allows the materialized content to be more granular as well as being able to "mix and match" results from different projects. eg you can get a 80% cache hit rate if only a small set of the code changed, or you could get cache hits from multiple builds if, for example, you cherry-picked 2 different changes. Additionally, the entire universe of build outputs doesn't need to be pulled down to the machine or compromises as to what specific "version" of a file is pulled down to the machine.

I'm not sure what you're referring to wrt something or other being a priority for build tools. I don't think that's relevant to the discussion for MSBuildCache, but I will say that .NET is open source so if this is a problem you feel is solvable, you can certainly work to help make that a reality. That and MSBuildCache basically _is_ this hash-based caching you're referring to anyway.

**Comment by fardarter** [comment]
@dfederm Let's leave aside the disk snapshotting and cache task ideas. I just mentioned them to explain how one could move state between machines. The fundamental problem I'm asking is how do I move the accumulated information from machine A -> machine B. Does that clear it up?

Let's take your example. I may have just memorised some of the capital cities, so when you ask me about them, I can respond quickly. Something in my brain must have stored that information -- the state of my brain has altered. That's why, when I go to sleep and wake up, I can still access the new information that I memorised.

Now imagine you did a Star Trek transportation on me. I would be reconstructed somewhere else, and the reconstructed me would also be able to recover the memory quickly. The transporter must know the "recipe" for that state. What info and relationships would let the new me boot up with all my memories in place.

So, the question is, when I transport my build process to a new machine, how do I reconstruct the state of things such that the cache "just works" and I don't need a new slow first build but can kick off as if I'd already done one?

Because if it is blown away every time I acquire a new machine on CI, then it seems like my only option to make it work is to have a persistent build agent. But I don't understand what mechanism would enforce this requirement on me -- if the required information for me to reconstruct the situation on a new machine is somewhere, then where is it? What's that recipe? I don't imagine you're storing it in RAM.

And if it wasn't reproducible, then what value would be added with the Azure Blob Storage implementation?

I don't know a huge amount about the details of caching, but I'd imagine you have a hashmap or some analogous data structure somewhere and then when you encounter a file, you hash the file and ask the hashmap if it has ever seen this? If it has, then it provides you either with the location of the file, and if it is remotely stored, you download it, and if it is in the correct local location, you just say "well, I got nothing to do here now -- don't process this". Am I way off?

> I'm not sure what you're referring to wrt something or other being a priority for build tools. I don't think that's relevant to the discussion for MSBuildCache, but I will say that .NET is open source so if this is a problem you feel is solvable, you can certainly work to help make that a reality. That and MSBuildCache basically is this hash-based caching you're referring to anyway.

I get that it is open source, but I'd still consider MS the vendor for .NET given the degree to which MS provides technical support. But I do take your point re open source.

**Comment by dfederm** [comment]
> The fundamental problem I'm asking is how do I move the accumulated information from machine A -> machine B.

_You_ don't. MSBuildCache takes care of uploading and downloading cache entries and content. That's the reason there are multiple implementations; each knows how to talk to a specific storage backend.

For the analogy we're using, there is no Star Trek transportation. I'm not bringing you with me, I'm just communicating with you. ie the cache is a service, not something local on the machine.

It might be helpful to illustrate what actually happens with the MSBuild cache plugin model and how the MSBuildCache uses it. So a build request is about to be queued and it first asks the plugin whether it wants to handle it. The plugin (MSBuildCache) computes a cache fingerprint based on the state of the repo in the context if the project being built and asks the Azure Pipeline cache service to **fetch the cache entry and content** with that fingerprint. If it's a cache hit, MSBuildCache will lay out the content on disk appropriately to mimic a "real" build.

So upon there being cache hits, the MSBuildCache itself is what is downloading the files from the relevant backend (Azure Pipeline cache in this case) and only for the projects which are cache hits.

> So, the question is, when I transport my build process to a new machine, how do I reconstruct the state of things such that the cache "just works" and I don't need a new slow first build but can kick off as if I'd already done one?

Just to drive this home, _you_ don't. MSBuildCache does this.

**Comment by fardarter** [comment]
@dfederm Thanks for being clear.

Reading back, I think the discussion on `git clean` and the local cache misled me into thinking this was the main cache. I understand now what is happening.
